docs/cuda_parallel/developer_overview.rst

```diff
@@ -0,0 +1,404 @@
+
+                extern "C" __global__ void kernel(storage_t value) {
+                    printf("thread %d prints value %d\n", threadIdx.x, op(value.data));
+                }
+            )XXX";
+
+            // ...
+            void *kernel_args[] = { value_ptr };
+            cuLaunchKernel((CUfunction)kernel, 1, 1, 1, 4, 1, 1, 0, 0, kernel_args, nullptr);
+
+The operator now takes a type-erased pointer.
+Let's take a look at the Python side:
+
+.. code-block:: python
+
+        import ctypes
+        import numba
+        import numba.cuda
+        import numpy as np
+
+        def op(value):
+            return numba.int32(value[0].real + value[0].imag)
+
+        value_type = numba.complex128
+        context = numba.cuda.descriptor.cuda_target.target_context
+        size = context.get_value_type(value_type).get_abi_size(context.target_data)
+        alignment = context.get_value_type(value_type).get_abi_alignment(context.target_data)
+        ltoir, _ = numba.cuda.compile(op, sig=numba.int32(numba.types.CPointer(value_type)), output='ltoir')
+
+        value = np.array([1 + 2j], dtype=np.complex128)
+        type_erased_value_ptr = value.ctypes.data_as(ctypes.c_void_p)
+
+        bindings = ctypes.CDLL('./build/libkernel.so')
+        bindings.launcher.argtypes = [ctypes.c_void_p, ctypes.c_int, ctypes.c_int, ctypes.c_char_p, ctypes.c_int]
+        bindings.launcher(type_erased_value_ptr, size, alignment, ltoir, len(ltoir))
+
+
+In the code above, we retreive type information (size and alignemnt) from the numba type system.
+One of the problems is that ``cuLaunchKernel`` accepts an array of pointers to CPU memory,
+from which it copies parameters for subsequent kernel launch.
+To get the pointer to CPU memory, we could use ``ctypes.byref``.
+Alternatively,
+we could allocate memory using ``numpy.array`` and later retreive the pointer with ``.ctypes.data_as(ctypes.c_void_p)``.
+
+The only missing part that separates us from cuda.parallel
+is the fact that all the kernels in CUDA C++ Core Compute Libraries are templates.
+Let's make our kernel a template as well.
+
+.. code-block:: c++
+
+    std::string kernel_source = storage_t + R"XXX(
+        extern "C" __device__ int op(char *state);
+
+        template <class T>
+        __global__ void kernel(T value) {
+            printf("thread %d prints value %d\n", threadIdx.x, op(value.data));
+        }
+    )XXX";
+
+
+Unfortunately, this is not sufficient.
+We have to instantiate the kernel template.
+To do that, we can use the following NVRTC API:
+
+
+.. code-block:: c++
+
+    nvrtcProgram prog;
+    const char *name = "test_kernel";
+    nvrtcCreateProgram(&prog, kernel_source.c_str(), name, 0, nullptr, nullptr);
+
+    // Get the name of the storage_t type
+    std::string storage_t_name;
+    nvrtcGetTypeName<storage_t>(&storage_t_name);
+
+    // Get the name of the kernel
+    std::string kernel_name = "kernel<" + storage_t_name + ">";
```

Comment:
This was discussed in a call a while ago. Now that we have this write-up (many thanks!), I see why it was argued that `nvrtcGetTypeName` is needed 😅  

This discussion presents a difference in the developer worldview: 
- For developers assembling the kernel strings in C++:
   - It is natural to say "I need to call `nvrtcGetTypeName` to get the type name so that I know what template specializations I need to feed it into `nvrtcAddNameExpression`". However, 
- For developers assembling the kernel strings in Python:
   - It is natural for us to think along the line that, since both `storage_t` and `kernel` strings are at our disposal, we can immediately assemble `kernel_name` in process and bypass `nvrtcGetTypeName`" (which is actually a [known trick](https://github.com/cupy/cupy/blob/b73ba5352850a84d1a86bcdad127619fee892fbc/docs/source/user_guide/kernel.rst?plain=1#L439)). This certainly requires "thinking everything at run time."

Below is a `cuda.core` based demo:
```python
from cuda.core.experimental import Device
from cuda.core.experimental import LaunchConfig, launch
from cuda.core.experimental import Program


code = """
/* these should come from libcudacxx, but I am lazy to set up include paths... */
template<class T, T v>
struct integral_constant { static constexpr T value = v; };
typedef integral_constant<bool, true> true_type;
typedef integral_constant<bool, false> false_type;

template<class T, class U>
struct is_same : false_type {};
 
template<class T>
struct is_same<T, T> : true_type {};

template< class T, class U >
constexpr bool is_same_v = is_same<T, U>::value;
/* *************************************************************************** */


// the point of this demo is: custom struct names can be easily referenced, and no
// need to call nvrtcGetTypeName on host (for which it's not possible to offer a
// Python binding anyway).
struct my_struct {
    int a;
    double b;
};


template<typename T>
__global__ void check_type() {
    // no RTTI on device, so let's do some tricks for the demo
    if constexpr (is_same_v<T, float>) {
        printf("it's float!\\n");
    }
    if constexpr (is_same_v<T, int>) {
        printf("it's int!\\n");
    }
    if constexpr (is_same_v<T, my_struct>) {
        printf("it's my_struct!\\n");
    }
}
"""


dev = Device()
dev.set_current()
arch = "sm_" + "".join(f"{i}" for i in dev.compute_capability)
s = dev.create_stream()

# prepare program
prog = Program(code, code_type="c++")
types = ("float", "int", "my_struct")
mod = prog.compile("cubin",
                   options=("-std=c++17", f"-arch={arch}",),
                   name_expressions=list(f"check_type<{t}>" for t in types))

# prepare launch config
config = LaunchConfig(grid=1, block=1, stream=s)

for t in types:
    # get kernel
    ker = mod.get_kernel(f"check_type<{t}>")
    
    # launch kernel on stream s
    launch(ker, config)
    s.sync()
```
Output:
```
it's float!
it's int!
it's my_struct!
```

One can say, "hey this is neat and we can do it in C++ too! It's not a Python-only thing!" That is certainly true, but it requires a change of the mindset to see this, which could take time to realize.